libvast/src/evaluate.cpp

```diff
@@ -0,0 +1,465 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2022 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "vast/arrow_table_slice.hpp"
+#include "vast/bitmap_algorithms.hpp"
+#include "vast/detail/assert.hpp"
+#include "vast/detail/overload.hpp"
+#include "vast/detail/passthrough.hpp"
+#include "vast/expression.hpp"
+#include "vast/ids.hpp"
+#include "vast/logger.hpp"
+#include "vast/table_slice.hpp"
+#include "vast/type.hpp"
+
+#include <arrow/record_batch.h>
+
+#include <cstddef>
+#include <span>
+
+namespace vast {
+
+namespace {
+
+// A utility function for evaluating meta extractors in predicates. This is
+// always a yes or no question per batch, so the function does not have to deal
+// with bitmaps at all.
+bool evaluate_meta_extractor(const table_slice& slice,
+                             const meta_extractor& lhs, relational_operator op,
+                             const data& rhs) {
+  switch (lhs.kind) {
+    case meta_extractor::kind::type:
+      return evaluate(materialize(slice.layout().name()), op, rhs);
+    case meta_extractor::kind::field: {
+      const auto* s = caf::get_if<std::string>(&rhs);
+      if (!s) {
+        VAST_WARN("#field can only compare with string");
```

Comment:
Why not pattern?

Reply:
@tobim IIRC you wrote this. Can't we just dispatch to the new `evaluate` here?

Reply:
That's actually quite implicit. Shouldn't `#field == /.*name/` do a suffix match on a fully qualified name?

(We'll create a story for this once we agree.)

Reply:
That's a totally different semantic, because `zeek.conn.id.orig_h` will be matched by `#field == "X"` for any of `orig_h`, `id.orig_h`, `zeek.conn.orig_h`, but not by `_h`.

Reply:
In my opinion, this operator is rather useless: `#field == "X"` is just the same as `X != nil`. We don't use it anywhere, really, so I'd rather just remove it.

Reply:
Let's pencil the removal for a separate story. I do see some value in it, but the discussion goes beyond the scope of this PR. It should probably also something we discuss in the community Slack.